### ISSUE TITLE
Possible bug in MPI_Put/Get

### ISSUE THREAD

**Issue by massimiliano-leoni** [pmodels/mpich#5879]
Hello, I have a piece of code like
```
MPI_Win_create
MPI_Win_fence

for (int i = 0; i < nProcs; ++i)
{
  MPI_Put // destination = i
}

MPI_Win_fence
MPI_Win_free

// Do stuff

MPI_Win_create
MPI_Win_fence

for (int i = 0; i < nProcs; ++i)
{
  MPI_Get/MPI_Put // source/destination = i
}

MPI_Win_fence
MPI_Win_free

```
in which I first distribute some data with `MPI_Put`, then do some work on it and finally send it back with either `MPI_Get` or `MPI_Put`.

This code works fine on my Ubuntu 20.04-based OS with OpenMPI 4.0.3, but it doesn't work on a CI bot running Ubuntu 20.04 with MPICH 4.0.
In particular, the first loop works fine: I compared rank-wise the data received after it and it's identical in OpenMPI and MPICH. The second loop is where dragons be: if I compare rank-wise the data that is transferred, the MPICH version has some nans and garbage numbers scattered in it.

My first version of this piece of code had an `MPI_Get` in the second loop, and as I realised that's where the issue was I replaced it with an equivalent `MPI_Put` but, to my great surprise, it still doesn't work!
Interestingly, with the `MPI_Get` version rank 0 was getting the correct data, all other ranks weren't.

I know this is a travesty of a decent bug report, but this happens deep down in a Finite Elements library and I'm unsure how to provide an MWE, or what else would be useful. I thought you'd know better what more information I can provide so please go ahead and ask for it.

I can provide a script that reproduces the bug using a docker image, if that is useful at all.

**Comment by hzhou** [comment]
Yes, please provide the reproducer with docker image. Meanwhile, please continue trying to get a minimal reproducer.

**Comment by massimiliano-leoni** [comment]
Ok!
Here is how to reproduce with a docker image.
Download the relevant code and the docker image with
```
git clone -b mleoni/interpolation https://github.com/massimiliano-leoni/dolfinx.git
cd dolfinx
sudo docker run -ti -p 127.0.0.1:8000:8000 -v $(pwd):/home/fenics/shared -w /home/fenics/shared fenicsproject/test-env:latest-mpich
```
Then, inside the docker image, paste
```
export DEBIAN_FRONTEND=noninteractive
export PETSC_ARCH=linux-gnu-real-32

# build dependencies
git clone https://github.com/FEniCS/basix.git --branch main --single-branch
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -B build-dir -S ./basix/cpp
cmake --build build-dir --parallel 12
cmake --install build-dir
pip3 install ./basix/python
pip3 install git+https://github.com/FEniCS/ufl.git
pip3 install git+https://github.com/FEniCS/ffcx.git
rm -rf basix build-dir

# build library
mkdir -p build
cd build
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ../cpp/
ninja -j12 install

# build tests
cd test/unit
cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug .
ninja -j12

# run test
mpirun -n 4 ./unittests [interpolation_different_meshes]
```
Once it has run, you'll see some output like
```
[RANK 1          ]      interpolation.cpp:84     ERR| diffNorm = inf
[RANK 3          ]      interpolation.cpp:84     ERR| diffNorm = inf
[RANK 0          ]      interpolation.cpp:84     ERR| diffNorm = 9.56855e-15
[RANK 2          ]      interpolation.cpp:84     ERR| diffNorm = -nan
```
where all the numbers should be very small but only rank 0 works fine.

Now, an explanation of what's happening. The code that runs is in `dolfinx/cpp/test/unit/interpolation.cpp`.
Line 62,
```
uR->interpolate(*uL);
```
is the interesting one and it ultimately calls the code at `dolfinx/cpp/dolfinx/fem/interpolate.h`, in particular the bit from line 643 to line 861. It looks like a lot but it really isn't :D 
Up to line 690 it computes, for each rank, a vector of `int` called `nPointsToSend` of size `nProcs` [size of communicator].
Then there is an `MPI_Allgather` that shares these vectors.

Then it computes offsets, allocates memory and creates the first `MPI_Win`. Then there's the loop with `MPI_Put`. Everything good up to here.

Lines 740-800 compute the array `values` from which other processes will read with `MPI_Get`, plus some offsets.
Then the code creates a window on `values` and every process collects from the `values` on other processes into its own `valuesToRetrieve` [the second loop].

The symptom is that the values that end up in `valuesToRetrieve` are wrong, usually the contain some garbage and/or nans.

Wherever you see the template parameter `T` you can just read `double` and `dolfinx::MPI::mpi_type<T>()` simply resolves to `MPI_DOUBLE`.

Please don't hesitate to let me know if I can provide any more information.

**Comment by massimiliano-leoni** [comment]
Never mind, it's all my fault because I was using `sizeof(MPI_DOUBLE)` and the like instead of `MPI_Type_size`. Sorry for the mess!

**Comment by hzhou** [comment]
No worries :) . I am glad you resolved the issue.

**Comment by hzhou** [comment]
> Never mind, it's all my fault because I was using `sizeof(MPI_DOUBLE)` and the like instead of `MPI_Type_size`. Sorry for the mess!

In case you wonder how it worked with OpenMPI. The datatype handle is 4-byte integer in MPICH, but a pointer type in OpenMPI.

**Comment by massimiliano-leoni** [comment]
Thanks! I am learning this right now on my skin :D
Long live continuous integration!
